src/ripple/app/ledger/TxHold.cpp

```diff
@@ -0,0 +1,727 @@
+#include <ripple/app/ledger/LedgerConsensus.h>
+#include <numeric>
+
+namespace ripple {
+
+class CandidateTxn
+{
+    friend class TxQImpl;
+
+protected:
+    boost::intrusive::set_member_hook<> byFee_;
+
+private:
+    STTx::pointer txn_;
+
+     uint64_t      feeLevel_;
+     uint32_t      lastRelayed_;
+     uint256       txID_;
+     boost::optional<uint256> priorTxID_;
+     Account       account_;
+     boost::optional<uint32_t> lastValid_;
+     uint32_t      sequence_;
+     TransactionEngineParams transactionParams_;
+
+public:
+    CandidateTxn (STTx::ref, TransactionEngine& engine,
+        std::uint32_t loadBase, TransactionEngineParams params);
+
+    STTx::pointer getTransaction()
+    {
+        return txn_;
+    }
+
+    uint64_t getFeeLevel() const
+    {
+        return feeLevel_;
+    }
+
+    uint32_t getSequence() const
+    {
+        return sequence_;
+    }
+};
+
+CandidateTxn::CandidateTxn(
+    STTx::ref txn,
+    TransactionEngine& engine,
+    std::uint32_t loadBase,
+    TransactionEngineParams params)
+    : txn_(txn)
+    , lastRelayed_(0), txID_(txn->getTransactionID())
+    , account_(txn->getSourceAccount().getAccountID())
+    , sequence_(txn->getSequence())
+    , transactionParams_(params)
+{
+    feeLevel_ = txn->getFeeLevelPaid(loadBase, engine.getLedger()->getBaseFee());
+
+    if (txn->isFieldPresent(sfLastLedgerSequence))
+    {
+        lastValid_ = txn->getFieldU32(sfLastLedgerSequence);
+    }
+
+    if (txn->isFieldPresent(sfAccountTxnID))
+    {
+        priorTxID_ = txn->getFieldH256(sfAccountTxnID);
+    }
+}
+
+class TxQAccount
+{
+    friend class TxQImpl;
+
+protected:
+    //boost::intrusive::set_member_hook<> byAccount_;
+
+private:
+
+    Account                    account_;
+    //uint64_t                   feeLevel_;
+    uint64_t                   totalFees_;
+    // Sequence number will be used as the key.
+    std::map <uint32_t, CandidateTxn> transactions_;
+
+public:
+    TxQAccount(STTx::ref txn);
+    TxQAccount(const Account& account);
+
+    //int getFeeLevel() const
+    //{
+    //    return feeLevel_;
+    //}
+
+    int getTxnCount() const
+    {
+        return transactions_.size();
+    }
+
+    bool empty() const
+    {
+        return !getTxnCount();
+    }
+
+    // Returns true if this is the new head transaction
+    // for this account
+    CandidateTxn&  addCandidate (CandidateTxn&&);
+    bool removeCandidate(uint32_t const&);
+
+    std::pair<bool, CandidateTxn const*> findCandidateAt(uint32_t const&);
+};
+
+TxQAccount::TxQAccount(STTx::ref txn)
+    :TxQAccount(txn->getFieldAccount160(sfAccount))
+{
+}
+
+TxQAccount::TxQAccount(const Account& account)
+    : account_(account)
+    //, feeLevel_(0)
+    , totalFees_(0)
+{ 
+}
+
+CandidateTxn&
+TxQAccount::addCandidate(CandidateTxn&& txn)
+{
+    auto fee = txn.getFeeLevel();
+    totalFees_ += fee;
+    auto sequence = txn.getSequence();
+
+    transactions_.emplace(std::make_pair(sequence, std::move(txn)));
```

Comment:
FWIW, I believe `emplace` returns an iterator to the position of the new element.  That could save you a couple of calls to `at()`.